compiler/rustc_transmute/src/layout/tree.rs

```diff
@@ -86,17 +87,18 @@ where
         F: Fn(D) -> bool,
     {
         match self {
-            Self::Seq(elts) => elts
-                .into_iter()
-                .map(|elt| elt.prune(f))
-                .try_fold(Tree::unit(), |elts, elt| {
+            Self::Seq(elts) => match elts.into_iter().map(|elt| elt.prune(f)).try_fold(
+                Tree::unit(),
+                |elts, elt| {
                     if elt == Tree::uninhabited() {
-                        Err(Tree::uninhabited())
+                        ControlFlow::Break(Tree::uninhabited())
                     } else {
-                        Ok(elts.then(elt))
+                        ControlFlow::Continue(elts.then(elt))
                     }
-                })
-                .into_ok_or_err(),
+                },
+            ) {
+                ControlFlow::Break(node) | ControlFlow::Continue(node) => node,
```

Comment:
I am more sympathetic to a `ControlFlow<T, T> -> T` conversion method than I am to `Result<T, T> -> T`. It's a bad sign that the original code here was an abuse of `Result` to begin with.

FWIW in my own code this would be:

```rust
    Self::Seq(elts) => {
        let mut pruned = Tree::unit();
        for elt in elts {
            pruned = match elt.prune(f) {
                uninhabited if uninhabited == Tree::uninhabited() => {
                    return Tree::uninhabited();
                }
                inhabited => pruned.then(inhabited),
            };
        }
        pruned
    }
```